File: hl_navigation_demos/src/sim.cpp
/**
 * @author Jerome Guzzi - <jerome@idsia.ch>
 */

#include <chrono>
#include <memory>
#include <vector>

#include "hl_navigation/kinematic.h"
#include "hl_navigation_sim/simulation.h"

using hl_navigation::TwoWheeled;
using hl_navigation::Vector2;
using hl_navigation_sim::BoundedStateEstimation;
using hl_navigation_sim::Experiment;
using hl_navigation_sim::WayPointsTask;
using hl_navigation_sim::World;

class ThymioDemo : public Experiment {
 public:
  explicit ThymioDemo(const char* behavior = "HL")
      : Experiment(0.02, 50 * 60), behavior(behavior) {}

 protected:
  void init(World& world, int seed) override {
    const std::vector<Vector2> targets{{1.0f, 0.0f}, {-1.0f, 0.0f}};
    for (size_t i = 0; i < 2; i++) {
      auto task = std::make_shared<WayPointsTask>(targets, true, 0.2);
      auto se = std::make_shared<BoundedStateEstimation>(&world, 1.0, 1.0);
      auto kinematic = std::make_shared<TwoWheeled>(0.166, 0.094);
      auto& agent = world.agents.emplace_back(behavior, 0.1, 0.08, kinematic,
                                              task, se, 0.02);
      agent.nav_behavior->set_optimal_speed(0.12);
      agent.nav_behavior->set_horizon(1.0);
      agent.nav_behavior->set_safety_margin(0.02);
      agent.nav_controller.set_speed_tolerance(0.01);
      agent.pose = {{i ? -0.5f : 0.5f, 0.5f}, 0.0f};
    }
    world.obstacles.emplace_back(Vector2{0.0f, 0.0f}, 0.1f);
  }

 private:
  const char* behavior;
};

static void show_usage(std::string name) {
  std::cout << "Usage: " << name << " <option(s)>" << std::endl
            << "Options:" << std::endl
            << "  --help\t\t\tShow this help message" << std::endl
            << "  --behavior=<BEHAVIOR>\tNavigation behavior" << std::endl;
}

int main(int argc, char* argv[]) {
  char behavior_name[10] = "HL";
  for (int i = 0; i < argc; i++) {
    if (sscanf(argv[i], "--behavior=%10s", behavior_name)) {
      continue;
    } else if (sscanf(argv[i], "--help")) {
      show_usage(argv[0]);
      return 0;
    }
  }
  ThymioDemo demo(behavior_name);
  printf("Start simulating 1 minute at 50 ticks per second\n");
  auto begin = std::chrono::high_resolution_clock::now();
  demo.run(0);
  auto end = std::chrono::high_resolution_clock::now();
  unsigned us =
      std::chrono::duration_cast<std::chrono::nanoseconds>(end - begin).count();
  printf("Done simulating in %.1f ms\n", us * 1e-6);
  return 0;
}

File: hl_navigation_demos/hl_navigation_demos/sim.py
import argparse
import time

import hl_navigation as nav
import hl_navigation_sim as sim


class ThymioDemo(sim.Scenario, name="PyThymioDemo"):  # type: ignore[call-arg]

    def __init__(self, behavior_type: str = "HL"):
        super().__init__()
        self._behavior_type = behavior_type

    def init_world(self, world: sim.World) -> None:
        targets = [(1.0, 0.0), (-1.0, 0.0)]
        for i in range(2):
            task = sim.WayPointsTask(targets, True, 0.2)
            se = sim.BoundedStateEstimation(world, 1.0, 1.0)
            kinematic = nav.kinematics.TwoWheeled(0.166, 0.094)
            behavior = nav.Behavior.make_type(self.behavior_type)
            agent = sim.Agent(0.08, behavior, kinematic, task, se, 0.02)
            agent.nav_behavior.optimal_speed = 0.12
            agent.nav_behavior.horizon = 1.0
            agent.nav_behavior.safety_margin = 0.02
            agent.nav_controller.speed_tolerance = 0.01
            agent.pose = nav.Pose2((-0.5 if i else 0.5, 0.5), 0.0)
            world.add_agent(agent)
        world.add_obstacle(nav.Disc((0.0, 0.0), 0.1))

    @sim.register_property("HL", "Behavior name")
    def behavior_type(self) -> str:
        return self._behavior_type

    @behavior_type.setter  # type: ignore[no-redef]
    def behavior_type(self, value: str) -> None:
        self._behavior_type = value


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--behavior", help="", type=str, default="HL")
    arg = parser.parse_args()
    demo = sim.Experiment(0.02, 50 * 60)
    demo.scenario = ThymioDemo(arg.behavior)
    demo.save_directory = "."
    demo.record_pose = True
    demo.name = "PyThymioDemo";
    print("Start simulating 1 minute at 50 ticks per second")
    begin = time.time()
    demo.run()
    end = time.time()
    ms = (end - begin) * 1e3
    print(f"Done simulating in {ms:.1f} ms")